### ISSUE TITLE
Are the build instructions in READER.md still valid?

### ISSUE THREAD

**Issue by tigran123** [koreader/koreader#2188]
This is just a question. It's been a very long time since I was hacking Koreader (in fact it was called "kindlepdfviewer" then :) but as I recently moved to the latest version of Ubuntu I may consider doing it again because it is much easier to do so than in Fedora.

Hence my question: are the build instructions in the `READER.md` still valid?

https://github.com/koreader/koreader#building-prerequisites

Specifically, is anyone using Ubuntu 16.04 (x86_64) for development and are there any specific build issues that I should be aware of?

**Comment by tigran123** [comment]
Yes, the instructions are still valid and the emulator builds fine on Ubuntu 16.04 without any glitches, so I am closing this issue.

**Comment by tigran123** [comment]
I closed this too soon, the build instructions are missing something because the compile for kobo target fails like this:


```
BUILD    willuslib/math.o
BUILD    willuslib/winshell.o
BUILD    willuslib/ocrtess.o
BUILD    willuslib/linux.o
BUILD    willuslib/wgui.o
BUILD    willuslib/point2d.o
BUILD    willuslib/willusversion.o
BUILD    willuslib/token.o
BUILD    willuslib/strbuf.o
BUILD    willuslib/fontdata.o
BUILD    willuslib/wincomdlg.o
BUILD    willuslib/winshellwapi.o
BUILD    willuslib/wgs.o
BUILD    willuslib/string.o
BUILD    willuslib/render.o
BUILD    willuslib/array.o
BUILD    willuslib/bmpdjvu.o
Makefile:83: recipe for target 'liblept.so' failed
make[5]: *** [liblept.so] Error 2
make[5]: Leaving directory '/home/tigran/koreader/base/thirdparty/libk2pdfopt/build/arm-linux-gnueabihf/libk2pdfopt-prefix/src/libk2pdfopt'
CMakeFiles/libk2pdfopt.dir/build.make:110: recipe for target 'libk2pdfopt-prefix/src/libk2pdfopt-stamp/libk2pdfopt-build' failed
make[4]: *** [libk2pdfopt-prefix/src/libk2pdfopt-stamp/libk2pdfopt-build] Error 2
make[4]: Leaving directory '/home/tigran/koreader/base/thirdparty/libk2pdfopt/build/arm-linux-gnueabihf'
CMakeFiles/Makefile2:67: recipe for target 'CMakeFiles/libk2pdfopt.dir/all' failed
make[3]: *** [CMakeFiles/libk2pdfopt.dir/all] Error 2
make[3]: Leaving directory '/home/tigran/koreader/base/thirdparty/libk2pdfopt/build/arm-linux-gnueabihf'
Makefile:83: recipe for target 'all' failed
make[2]: *** [all] Error 2
make[2]: Leaving directory '/home/tigran/koreader/base/thirdparty/libk2pdfopt/build/arm-linux-gnueabihf'
Makefile.third:185: recipe for target 'build/arm-linux-gnueabihf/libs/libk2pdfopt.so.2' failed
make[1]: *** [build/arm-linux-gnueabihf/libs/libk2pdfopt.so.2] Error 2
make[1]: Leaving directory '/home/tigran/koreader/base'
Makefile:91: recipe for target 'base/build/arm-linux-gnueabihf/luajit' failed
make: *** [base/build/arm-linux-gnueabihf/luajit] Error 2
```

**Comment by tigran123** [comment]
Rerunning compile shows this:

```
$ ./kodev release kobo
make -C base
make[1]: Entering directory '/altroot/home/tigran/koreader/base'
install -d thirdparty/libk2pdfopt/build/arm-linux-gnueabihf
cd thirdparty/libk2pdfopt/build/arm-linux-gnueabihf && \
	cmake -DHOST="arm-linux-gnueabihf" \
	-DCC="/usr/bin/ccache arm-linux-gnueabihf-gcc -static-libstdc++" -DCFLAGS="-O2 -ffast-math -pipe -fomit-frame-pointer  -march=armv7-a -mtune=cortex-a8 -mfpu=neon -mthumb -mfloat-abi=hard -fno-finite-math-only -fno-stack-protector -U_FORTIFY_SOURCE -fPIC -DKOBO_PLATFORM" -DCXX="/usr/bin/ccache arm-linux-gnueabihf-g++ -static-libstdc++" -DCXXFLAGS="-O2 -ffast-math -pipe -fomit-frame-pointer  -march=armv7-a -mtune=cortex-a8 -mfpu=neon -mthumb -mfloat-abi=hard -fno-finite-math-only -fno-stack-protector -U_FORTIFY_SOURCE -fPIC -O3" \
	-DAR="/usr/bin/ccache arm-linux-gnueabihf-ar" -DSTDCPPLIB="/usr/lib/gcc-cross/arm-linux-gnueabihf/5/libstdc++.a" -DLDFLAGS="-Wl,-O1 -Wl,--as-needed -static-libstdc++" \
	-DZLIB_DIR=/altroot/home/tigran/koreader/base/thirdparty/zlib/build/arm-linux-gnueabihf/zlib-prefix/src/zlib -DZLIB=/altroot/home/tigran/koreader/base/build/arm-linux-gnueabihf/libs/libz.so.1 -DPNG_DIR=/altroot/home/tigran/koreader/base/thirdparty/libpng/build/arm-linux-gnueabihf/libpng-prefix/src/libpng-build \
	-DLEPTONICA_DIR=/altroot/home/tigran/koreader/base/thirdparty/leptonica/build/arm-linux-gnueabihf/leptonica-prefix/src/leptonica -DTESSERACT_DIR=/altroot/home/tigran/koreader/base/thirdparty/tesseract/build/arm-linux-gnueabihf/tesseract-prefix/src/tesseract \
	/altroot/home/tigran/koreader/base/thirdparty/libk2pdfopt && \
	make
-- Configuring done
-- Generating done
-- Build files have been written to: /altroot/home/tigran/koreader/base/thirdparty/libk2pdfopt/build/arm-linux-gnueabihf
make[2]: Entering directory '/altroot/home/tigran/koreader/base/thirdparty/libk2pdfopt/build/arm-linux-gnueabihf'
make[3]: Entering directory '/altroot/home/tigran/koreader/base/thirdparty/libk2pdfopt/build/arm-linux-gnueabihf'
make[4]: Entering directory '/altroot/home/tigran/koreader/base/thirdparty/libk2pdfopt/build/arm-linux-gnueabihf'
make[4]: Leaving directory '/altroot/home/tigran/koreader/base/thirdparty/libk2pdfopt/build/arm-linux-gnueabihf'
make[4]: Entering directory '/altroot/home/tigran/koreader/base/thirdparty/libk2pdfopt/build/arm-linux-gnueabihf'
[ 12%] Performing build step for 'libk2pdfopt'
make[5]: Entering directory '/altroot/home/tigran/koreader/base/thirdparty/libk2pdfopt/build/arm-linux-gnueabihf/libk2pdfopt-prefix/src/libk2pdfopt'
configure: WARNING: if you wanted to set the --build type, don't use --host.
    If a cross compiler is detected then cross compile mode will be used
configure: WARNING: using cross tools not prefixed with host triplet
Makefile:83: recipe for target 'liblept.so' failed
make[5]: *** [liblept.so] Error 2
make[5]: Leaving directory '/altroot/home/tigran/koreader/base/thirdparty/libk2pdfopt/build/arm-linux-gnueabihf/libk2pdfopt-prefix/src/libk2pdfopt'
CMakeFiles/libk2pdfopt.dir/build.make:110: recipe for target 'libk2pdfopt-prefix/src/libk2pdfopt-stamp/libk2pdfopt-build' failed
make[4]: *** [libk2pdfopt-prefix/src/libk2pdfopt-stamp/libk2pdfopt-build] Error 2
make[4]: Leaving directory '/altroot/home/tigran/koreader/base/thirdparty/libk2pdfopt/build/arm-linux-gnueabihf'
CMakeFiles/Makefile2:67: recipe for target 'CMakeFiles/libk2pdfopt.dir/all' failed
make[3]: *** [CMakeFiles/libk2pdfopt.dir/all] Error 2
make[3]: Leaving directory '/altroot/home/tigran/koreader/base/thirdparty/libk2pdfopt/build/arm-linux-gnueabihf'
Makefile:83: recipe for target 'all' failed
make[2]: *** [all] Error 2
make[2]: Leaving directory '/altroot/home/tigran/koreader/base/thirdparty/libk2pdfopt/build/arm-linux-gnueabihf'
Makefile.third:185: recipe for target 'build/arm-linux-gnueabihf/libs/libk2pdfopt.so.2' failed
make[1]: *** [build/arm-linux-gnueabihf/libs/libk2pdfopt.so.2] Error 2
make[1]: Leaving directory '/altroot/home/tigran/koreader/base'
Makefile:52: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by NiLuJe** [comment]
Can't tell anything from those, we need verbose build logs... which are disabled to make Travis behave/readable.

I have a mess of hackish tweaks to bypass that, but some of it needs deeper integration in the build-system. Can't remember the details, but see if you get better output after this:

```
cd base
sed -e 's# --silent >/dev/null 2>&1##g' -i thirdparty/libk2pdfopt/build/git_checkout/libk2pdfopt/Makefile
sed -e 's# --silent##g' -i thirdparty/libpng/CMakeLists.txt thirdparty/tar/CMakeLists.txt thirdparty/zyre/CMakeLists.txt thirdparty/zsync/CMakeLists.txt thirdparty/luasec/CMakeLists.txt thirdparty/libzmq/CMakeLists.txt thirdparty/czmq/CMakeLists.txt thirdparty/zlib/CMakeLists.txt thirdparty/gettext/CMakeLists.txt thirdparty/freetype2/CMakeLists.txt thirdparty/libiconv/CMakeLists.txt thirdparty/djvulibre/CMakeLists.txt thirdparty/filemq/CMakeLists.txt thirdparty/openssl/CMakeLists.txt thirdparty/giflib/CMakeLists.txt thirdparty/libjpeg-turbo/CMakeLists.txt
sed -e 's# >/dev/null 2>&1##g' -i thirdparty/openssl/CMakeLists.txt
cd -
```

**Comment by tigran123** [comment]
@NiLuJe Thank you. With verbose logs we get:

```
/usr/lib/gcc-cross/arm-linux-gnueabihf/5/../../../../arm-linux-gnueabihf/bin/ld: .libs/liblept.so.4.0.3: No symbol version section for versioned symbol `fmemopen@GLIBC_2.2.5'
/usr/lib/gcc-cross/arm-linux-gnueabihf/5/../../../../arm-linux-gnueabihf/bin/ld: final link failed: Nonrepresentable section on output
collect2: error: ld returned 1 exit status
Makefile:437: recipe for target 'liblept.la' failed
make[7]: *** [liblept.la] Error 1
make[7]: Leaving directory '/altroot/home/tigran/koreader/base/thirdparty/leptonica/build/arm-linux-gnueabihf/leptonica-prefix/src/leptonica/src'
Making install in prog
make[7]: Entering directory '/altroot/home/tigran/koreader/base/thirdparty/leptonica/build/arm-linux-gnueabihf/leptonica-prefix/src/leptonica/prog'
make[7]: *** No rule to make target '../src/liblept.la', needed by 'convertfilestopdf'. Stop.
make[7]: *** Waiting for unfinished jobs....
  CC     convertfilestopdf.o
make[7]: Leaving directory '/altroot/home/tigran/koreader/base/thirdparty/leptonica/build/arm-linux-gnueabihf/leptonica-prefix/src/leptonica/prog'
make[7]: Entering directory '/altroot/home/tigran/koreader/base/thirdparty/leptonica/build/arm-linux-gnueabihf/leptonica-prefix/src/leptonica'
make[8]: Entering directory '/altroot/home/tigran/koreader/base/thirdparty/leptonica/build/arm-linux-gnueabihf/leptonica-prefix/src/leptonica'
 /usr/bin/install -c -m 644 lept.pc '/altroot/home/tigran/koreader/base/thirdparty/leptonica/build/arm-linux-gnueabihf/leptonica-prefix/src/leptonica/lib/pkgconfig'
make[8]: Leaving directory '/altroot/home/tigran/koreader/base/thirdparty/leptonica/build/arm-linux-gnueabihf/leptonica-prefix/src/leptonica'
make[7]: Leaving directory '/altroot/home/tigran/koreader/base/thirdparty/leptonica/build/arm-linux-gnueabihf/leptonica-prefix/src/leptonica'
Makefile:366: recipe for target 'install-recursive' failed
make[6]: *** [install-recursive] Error 1
make[6]: Leaving directory '/altroot/home/tigran/koreader/base/thirdparty/leptonica/build/arm-linux-gnueabihf/leptonica-prefix/src/leptonica'
Makefile:83: recipe for target 'liblept.so' failed
make[5]: *** [liblept.so] Error 2
make[5]: Leaving directory '/altroot/home/tigran/koreader/base/thirdparty/libk2pdfopt/build/arm-linux-gnueabihf/libk2pdfopt-prefix/src/libk2pdfopt'
CMakeFiles/libk2pdfopt.dir/build.make:110: recipe for target 'libk2pdfopt-prefix/src/libk2pdfopt-stamp/libk2pdfopt-build' failed
make[4]: *** [libk2pdfopt-prefix/src/libk2pdfopt-stamp/libk2pdfopt-build] Error 2
make[4]: Leaving directory '/altroot/home/tigran/koreader/base/thirdparty/libk2pdfopt/build/arm-linux-gnueabihf'
CMakeFiles/Makefile2:67: recipe for target 'CMakeFiles/libk2pdfopt.dir/all' failed
make[3]: *** [CMakeFiles/libk2pdfopt.dir/all] Error 2
make[3]: Leaving directory '/altroot/home/tigran/koreader/base/thirdparty/libk2pdfopt/build/arm-linux-gnueabihf'
Makefile:83: recipe for target 'all' failed
make[2]: *** [all] Error 2
make[2]: Leaving directory '/altroot/home/tigran/koreader/base/thirdparty/libk2pdfopt/build/arm-linux-gnueabihf'
Makefile.third:185: recipe for target 'build/arm-linux-gnueabihf/libs/libk2pdfopt.so.2' failed
make[1]: *** [build/arm-linux-gnueabihf/libs/libk2pdfopt.so.2] Error 2
make[1]: Leaving directory '/altroot/home/tigran/koreader/base'
Makefile:91: recipe for target 'base/build/arm-linux-gnueabihf/luajit' failed
make: *** [base/build/arm-linux-gnueabihf/luajit] Error 2
```

**Comment by tigran123** [comment]
For ARM toolchain I installed this as in the `README.md`:

```
sudo apt-get install gcc-arm-linux-gnueabihf g++-arm-linux-gnueabihf
```

**Comment by NiLuJe** [comment]
Right. That TC probably uses a much newer glibc, which would most likely cause some form of ABI mismatches in the resulting binaries when trying to run them on the Kobo, despite our best efforts.

But, more to the point, one of those efforts is the leptonica workaround we have in place, which, unlike the other *FLAGS based one, *requires* a TC glibc that's older than the fmemopen switch, or it fails to build just like that.

TL;DR: Use a proper TC that properly targets the Kobo's glibc.

**Comment by tigran123** [comment]
Well, I used the one that `README.md` pointed me to. Is there a different one I should use? In the old days of kindlepdfviewer development I used one of these:

```
$ l
total 135116
-rw-rw-r-- 1 tigran tigran 92613573 Dec  9  2013 arm-2012.03-57-arm-none-linux-gnueabi-i686-pc-linux-gnu.tar.bz2
-rw-r--r-- 1 tigran tigran 45741995 Dec  9  2013 arm-kindle-linux-x-tools-glibc2.5-gcc4.2.4.tar.gz
```

**Comment by NiLuJe** [comment]
Problem with the Ubuntu one is it's a moving target, it might have been a rough match at one point, but it may not be so anymore.

I've edited my previous answer with a couple queries to figure if at least the buildsystem properly detected it or not...

**Comment by NiLuJe** [comment]
Which Ubuntu build are you using?

I haver a sinking feeling their cross glibc has been mangled all to hell... It's currently using glibc 2.23, and the cutoff point for the fmemopen is glibc >= 2.22. We currently only check the version, not the actual symbol table (because backporting symbol table changes seems fairly insane), so if they for some reason reverted the fmemopen changes in their glibc build, our buildsystem assumes it's >= 2.22, so it enables the workaround to manually pick the old symbol. But if the change has been reverted, that symbol doesn't exist -> link failure.

**Comment by tigran123** [comment]
I am using Ubuntu 16.04. Here is some info:

```
$ cat hello.c
#include <stdio.h>

int main(int argc, char *argv[])
{
	printf("Hello, world\n");
	return 0;
}
$ arm-linux-gnueabihf-
arm-linux-gnueabihf-addr2line     arm-linux-gnueabihf-dwp           arm-linux-gnueabihf-gcc-ar        arm-linux-gnueabihf-gcov          arm-linux-gnueabihf-ld.bfd        arm-linux-gnueabihf-readelf
arm-linux-gnueabihf-ar            arm-linux-gnueabihf-elfedit       arm-linux-gnueabihf-gcc-ar-5      arm-linux-gnueabihf-gcov-5        arm-linux-gnueabihf-ld.gold       arm-linux-gnueabihf-size
arm-linux-gnueabihf-as            arm-linux-gnueabihf-g++           arm-linux-gnueabihf-gcc-nm        arm-linux-gnueabihf-gcov-tool     arm-linux-gnueabihf-nm            arm-linux-gnueabihf-strings
arm-linux-gnueabihf-c++filt       arm-linux-gnueabihf-g++-5         arm-linux-gnueabihf-gcc-nm-5      arm-linux-gnueabihf-gcov-tool-5   arm-linux-gnueabihf-objcopy       arm-linux-gnueabihf-strip
arm-linux-gnueabihf-cpp           arm-linux-gnueabihf-gcc           arm-linux-gnueabihf-gcc-ranlib    arm-linux-gnueabihf-gprof         arm-linux-gnueabihf-objdump       
arm-linux-gnueabihf-cpp-5         arm-linux-gnueabihf-gcc-5         arm-linux-gnueabihf-gcc-ranlib-5  arm-linux-gnueabihf-ld            arm-linux-gnueabihf-ranlib        
$ arm-linux-gnueabihf-gcc hello.c -o hello
$ file hello
hello: ELF 32-bit LSB executable, ARM, EABI5 version 1 (SYSV), dynamically linked, interpreter /lib/ld-linux-armhf.so.3, for GNU/Linux 3.2.0, BuildID[sha1]=21b03fa33229042b71290ef48739f5389831756a, not stripped
$ arm-linux-gnueabihf-objdump -R hello

hello:     file format elf32-littlearm

DYNAMIC RELOCATION RECORDS
OFFSET   TYPE              VALUE 
0002101c R_ARM_GLOB_DAT    __gmon_start__
0002100c R_ARM_JUMP_SLOT   puts@GLIBC_2.4
00021010 R_ARM_JUMP_SLOT   __libc_start_main@GLIBC_2.4
00021014 R_ARM_JUMP_SLOT   __gmon_start__
00021018 R_ARM_JUMP_SLOT   abort@GLIBC_2.4
```

**Comment by NiLuJe** [comment]
Not particularly insightful ;).

As mentioned earlier, what would be is this:

```realpath $(arm-linux-gnueabihf-gcc -print-file-name=libc.so.6)```

And

```readelf -s realpath $(arm-linux-gnueabihf-gcc -print-file-name=libc.so.6) | grep fmemopen```

**Comment by tigran123** [comment]
Ok:

```
$ realpath $(arm-linux-gnueabihf-gcc -print-file-name=libc.so.6)
/usr/arm-linux-gnueabihf/lib/libc-2.23.so
$ readelf -s realpath $(arm-linux-gnueabihf-gcc -print-file-name=libc.so.6) | grep fmemopen
readelf: Error: 'realpath': No such file
  1164: 0004b9d5   256 FUNC    GLOBAL DEFAULT   13 fmemopen@GLIBC_2.4
  1167: 0004b6f9   296 FUNC    GLOBAL DEFAULT   13 fmemopen@@GLIBC_2.22
```

**Comment by NiLuJe** [comment]
Yuuuup, they messed with the fmemopen symbols.

Yay'.

**Comment by NiLuJe** [comment]
Can you check if the native x86 one is mangled that way, too?

For ref, on my system:
```
  1173: 00000000000703f0   407 FUNC    GLOBAL DEFAULT   11 fmemopen@GLIBC_2.2.5
  1174: 0000000000070080   407 FUNC    GLOBAL DEFAULT   11 fmemopen@@GLIBC_2.22
```

**Comment by tigran123** [comment]
Here is the output on my system:
```
$ readelf -s $(gcc -print-file-name=libc.so.6) | grep fmemopen
  1174: 0000000000077c90   391 FUNC    GLOBAL DEFAULT   13 fmemopen@GLIBC_2.2.5
  1175: 0000000000077910   423 FUNC    GLOBAL DEFAULT   13 fmemopen@@GLIBC_2.22
```

**Comment by NiLuJe** [comment]
Yep, which is sane (https://sourceware.org/git/?p=glibc.git;a=commit;h=fdb7d390dd0d96e4a8239c46f3aa64598b90842b).

I have no idea what they did to keep the compat symbol versioned to @GLIBC_2.4 (which is actually how the symbol was verisoned before glibc 2.22, IIRC), but that's obviously unexpected, hence why the patch fails.

**Comment by tigran123** [comment]
So, should I edit one of those two leptonica patches to point to `fmemopen@GLIBC_2.4` or `fmemopen@GLIBC_2.22`?

**Comment by NiLuJe** [comment]
I'd go with use a sane TC instead, but, yeah, a workaround would be to tweak the patch to point to the @GLIBC_2.4 one.

That's how the symbol was versioned in glibc 2.15 (fmemopen@@GLIBC_2.4), so that *should* work on the Kobo.

**Comment by tigran123** [comment]
Ok, so where do I get a sane TC from?

**Comment by NiLuJe** [comment]
https://github.com/koreader/koxtoolchain ;)

**Comment by NiLuJe** [comment]
Aaaaaaggggh, dammit.

I know what went wrong.

The fmemopen compat symbol is pinned to a different version on arm than on x86.

So, that builod is actually sane, my bad.

**Comment by NiLuJe** [comment]
cf. https://sourceware.org/git/?p=glibc.git;a=blob;f=sysdeps/unix/sysv/linux/x86_64/64/libc.abilist;h=c6e3cd42572d7e2a6831e74213ea20f8bb918ccb;hb=refs/heads/release/2.23/master

vs.

https://sourceware.org/git/?p=glibc.git;a=blob;f=sysdeps/unix/sysv/linux/arm/libc.abilist;h=db9fa350f09e22da8af03bfa431f791000726d3a;hb=refs/heads/release/2.23/master

I'll fix the patch...

**Comment by tigran123** [comment]
Oh, excellent, so I'll await your commit. Thank you, brother! :)

**Comment by NiLuJe** [comment]
This is potentially a good news though, it makes things much less weird, that version mismatch between what's in the arm glibc-2.15 and that x86 compat symbol always irked me... ;).

**Comment by NiLuJe** [comment]
https://github.com/NiLuJe/koreader-base/commit/0778dc59efbfadd9dff0e2f3811f922b9fc492bc

Should do the trick.

Thanks for your quick answers, and sorry for the confusion, was definitely not expecting that the arm symbol could actually be pinned to a different version than on x86. >_<".

**Comment by tigran123** [comment]
Btw, trying that `koxtoolchain` failed as well:

```
[ALL  ]    {standard input}: Assembler messages:
[ALL  ]    {standard input}:98: Error: invalid swi expression
[ALL  ]    {standard input}:98: Error: value of 983045 too large for field of 2 bytes at 186
[ALL  ]    /home/tigran/Kobo/koxtoolchain/build/kobo/.build/arm-kobo-linux-gnueabihf/build/build-libc-startfiles/sysd-rules:3219: recipe for target '/home/tigran/Kobo/koxtoolchain/build/kobo/.build/arm-kobo-linux-gnueabihf/build/build-libc-startfiles/csu/libc-tls.o' failed
[ERROR]    make[3]: *** [/home/tigran/Kobo/koxtoolchain/build/kobo/.build/arm-kobo-linux-gnueabihf/build/build-libc-startfiles/csu/libc-tls.o] Error 1
[ALL  ]    make[3]: *** Waiting for unfinished jobs....
[ALL  ]    make[3]: Leaving directory '/home/tigran/Kobo/koxtoolchain/build/kobo/.build/src/glibc-2.11/csu'
[ALL  ]    Makefile:207: recipe for target 'csu/subdir_lib' failed
[ERROR]    make[2]: *** [csu/subdir_lib] Error 2
[ALL  ]    make[2]: Leaving directory '/home/tigran/Kobo/koxtoolchain/build/kobo/.build/src/glibc-2.11'
[ALL  ]    Makefile:12: recipe for target 'csu/subdir_lib' failed
[ERROR]    make[1]: *** [csu/subdir_lib] Error 2
[ALL  ]    make[1]: Leaving directory '/home/tigran/Kobo/koxtoolchain/build/kobo/.build/arm-kobo-linux-gnueabihf/build/build-libc-startfiles'
[ERROR]
[ERROR]  >>
[ERROR]  >>  Build failed in step 'Installing C library headers & start files'
[ERROR]  >>        called in step '(top-level)'
[ERROR]  >>
[ERROR]  >>  Error happened in: CT_DoExecLog[scripts/functions@257]
[ERROR]  >>        called from: do_libc_backend_once[scripts/build/libc/glibc.sh@450]
[ERROR]  >>        called from: do_libc_backend[scripts/build/libc/glibc.sh@153]
[ERROR]  >>        called from: do_libc_start_files[scripts/build/libc/glibc.sh@70]
[ERROR]  >>        called from: main[scripts/crosstool-NG.sh@646]
[ERROR]  >>
```

But hopefully we won't need this as your commit above should fix the problem. I'll test it later (I am now re-building koreader for emulator).

**Comment by NiLuJe** [comment]
@tigran123: Possibly because Ubuntu moved to GCC 6.1?

I don't think I backported the changes needed to handle that (mostly because I won't touch a GCC point release with a ten-foot pole, running a source distribution ;)).

**Comment by houqp** [comment]
Oh yes, I was only able to get the koxtoolchain kobo toolchain compiled properly under unbuntu 14.04. It starts to fail on arch and newer versions of ubuntu. I forgot where I got the 2.11 version, but I have definitely seen it somewhere before. Perhaps I also got it from the misleading filename, LOL.

Anyway, looks like we should bump the version for kobo then. Hopefully this resolves the issue.

**Comment by NiLuJe** [comment]
@houqp: The Arch issues might be more generalized if the static library snafu still hasn't been resolved.

But yeah, it definitely should be using glibc 2.15 ;).

**Comment by tigran123** [comment]
@houqp What does "milestone --- next stable release" mean? Do you mean to say it won't be fixed until _after_ the next stable release or that it will be fixed _before_ the next stable release? I think it is quite important to be able to build Koreader under Ubuntu 16.04. Anyway, in the worst case I can apply @NiLuJe 's patch manually, I just have to remember how to point to his git repository from mine (I remember the day when @dpavlin wrote those git instructions for me many many years ago :)

**Comment by houqp** [comment]
> @houqp: The Arch issues might be more generalized if the lack of static library thingy still hasn't been resolved. (i.e., ct-ng requires static libraries, and Arch stopped shipping them).

Oh interesting... I haven't thought of this before... What kind of static libraries are you referring to?

> @houqp What does "milestone --- next stable release" mean? Do you mean to say it won't be fixed until after the next stable release or that it will be fixed before the next stable release? I think it is quite important to be able to build Koreader under Ubuntu 16.04.

I means this needs to be fixed before we tag the next stable release. I put all the high priority issues in that milestone.

> Anyway, in the worst case I can apply @NiLuJe 's patch manually, I just have to remember how to point to his git repository from mine (I remember the day when @dpavlin wrote those git instructions for me many many years ago :)

@NiLuJe, you will merge the patch to koreader/koreader-base right?

**Comment by NiLuJe** [comment]
@houqp: cf. https://github.com/crosstool-ng/crosstool-ng/issues/297

I just sent the PR, but since I don't run any of the affected TCs, I can't check how well it actually behave on affected systems ;).

**Comment by houqp** [comment]
> I just sent the PR, but since I don't run any of the affected TCs, I can't check how well it actually behave on affected systems ;).

It's already broken, so it can't be worse :)

**Comment by NiLuJe** [comment]
True ;).

**Comment by tigran123** [comment]
Well, I promise to test it as soon as it is merged :)

**Comment by houqp** [comment]
merged, I also updated https://github.com/koreader/koxtoolchain to use glibc 2.15 for kobo. It should be working for 16.04 now :)

**Comment by NiLuJe** [comment]
I'll send a PR to update base in koreader ;).

Quick question: what's the best way to deal with squash-merged PRs in our own branches?

**Comment by tigran123** [comment]
@houqp You need also to merge the "Update base" PR so that I can see the change from a cloned "koreader" module.

**Comment by houqp** [comment]
> Quick question: what's the best way to deal with squash-merged PRs in our own branches?
> 
> (After being confused for a bit, I basically just did a rebase, dropping my non-squashed version of the commits, then a force push).

Sorry, my bad. After used the squash feature once, now github UI always default all my merge action to squash merge and I kept forgetting to change it back to normal merge. Unfortunately, you will have to discard the old commit. The squash merge only works well when you always delete the old branch and start a new one from master every time.

tldr; I fucked it up and will try to remember not using this confusing feature unless really necessary. What I really want is a rebase merge button that gitlab has >_<

**Comment by NiLuJe** [comment]
No worries ;)

**Comment by tigran123** [comment]
I confirm that Kobo package builds successfully, but  the binaries are completely different from the official ones:

```
$ find -name luajit -ls
  1781025    296 -rwxrwxr-x   1 tigran   tigran     301036 Jul 29 09:03 ./koreader/luajit
  1780220    288 -rwxr-xr-x   1 tigran   tigran     292900 Jul 27 23:18 ./koreader-official/luajit
$ file koreader/luajit koreader-official/luajit 
koreader/luajit:          ELF 32-bit LSB executable, ARM, EABI5 version 1 (SYSV), dynamically linked, interpreter /lib/ld-linux-armhf.so.3, for GNU/Linux 3.2.0, BuildID[sha1]=c212eed51a74b01fe0da376dd857fff8b91a25f9, stripped
koreader-official/luajit: ELF 32-bit LSB executable, ARM, EABI5 version 1 (SYSV), dynamically linked, interpreter /lib/ld-linux-armhf.so.3, for GNU/Linux 2.6.31, BuildID[sha1]=fd1b7782d724ae8030add284956ce6d6b42129c2, stripped
```

Anyway, the good news is that the binaries work perfectly fine! In fact, it feels slightly faster (turning the pages), but that is probably a purely psychological effect --- "things built with your own hands always feel better" :)

So, now we know that Koreader built with the latest TC under Ubuntu 16.04 (on x86_64 arch) works fine on Kobo Aura H2O.

**Comment by NiLuJe** [comment]
The move to GCC 5 may very well make things slightly snappier ;).

**Comment by tigran123** [comment]
On putting the device to sleep it crashed once, with this error:

```
glic: detected luajit free() invalid pointer 0x.....
```

This is the same error that I encountered when porting StarPlot to the latest Ubuntu caused by code like this:

```
sortable *temparray = new sortable[size];
....
delete [] temparray;
```

I am no C++ expert, but apparently (checking answers to this problem on stackoverflow) it is undefined behaviour to do this. On Fedora 17 (i.e. on much more ancient g++ and glibc combination) this worked fine, but on Ubuntu 16.04 this caused the same error message as above.

**Comment by NiLuJe** [comment]
@tigran123 : Might be relevant to #1974

**Comment by tigran123** [comment]
Btw, on my Kobo Aura H2O the file `/sys/power/wakeup_count` doesn't even exist.

Also, looking at CoolReader's `vlasovsoft/suspend.sh` script I see that it does something very different from `koreader/suspend.sh`, namely there is an extra command (to resume):

```
echo 0 > /sys/power/state-extended
```

and no reads or writes to `/sys/power/wakeup_count`. Plus there is a SIGTERM handler. And I never had any problem with its sleep/wakeup mechanism. But one cannot just modify `koreader/suspend.sh` to match `vlasovsoft/suspend.sh` because in Koreader there are separate calls `Kobo:suspend()` and `Kobo:resume()`. Why not do both suspend and resume in a single script `suspend.sh` like CoolReader does?

**Comment by chrox** [comment]
It's no shame to port CoolReader's suspend script to koreader with proper credit now that CR's script seems much better than koreader's regarding to suspending and resuming.

**Comment by NiLuJe** [comment]
@tigran123:

Setting state-extended to 0 on *suspend* is utterly nonsensical. Check the Kernel, it just sets a global (gSleep_Mode_Suspend) that's checked when going to suspend in order to suspend various subsystems.

Setting it to 0 on *wakeup*, though, yes, and that's what nickel does.

wakeup-count *definitely* exists on my H2O (running the latest FW).

I fail to see the interest of the SIGTERM handled since this script is fired as a one-off, and doesn't hang around in the background.

**Comment by tigran123** [comment]
@NiLuJe Is it really nonsensical? I thought that it was setting state-extended to 0 precisely for the reason you describe, namely on wakeup. I.e. the execution of the `suspend.sh` script is continued from the point "just before" writing 0, no? So, I assumed that `suspend.sh` script in CR does the job of both suspending and handling the wakeup in one place, instead of doing it in two different functions in Koreader. But in total Koreader does the same PLUS your code for writing 'a' to neocmd to really wake IR touch subsystem up.

Btw, in some issue someone commented that seeing "END:" in the log means suspend failed. No, it means the script resumed its normal execution after the system woke up. The timestamp in the log is erroneous/confusing because it is the same as the message "Asked to suspend to RAM..." but this message is only printed when the system is woken up. So, the two messages "Asked to suspend..." and "END!" are printed when you wake up, but all the other messages are printed when you suspend.

**Comment by houqp** [comment]
Just so you know. As @NiLuJe pointed out, there is a known mem leak issue that will cause luajit to segfault when the device is running out of memory. It usually happens on resume or during suspend.

So the segfault might have nothing to do with the suspend script. That said, it won't hurt to give coolreader's script a try.

**Comment by tigran123** [comment]
@houqp The issue I encountered was glibc detecting invalid pointer passed to free(). Anyway, I could not reproduce it even though I have done many suspend/wakeups yesterday and even left the device in Koreader's suspended state overnight. Everything was fine this morning --- even the battery was still at exactly 100%, so it really was asleep (otherwise at least 2-3 percent battery would have been eaten). So, at the moment everything seems fine with Koreader's suspend.sh (ah, yes, I removed all that wakeup_count stuff from `suspend.sh` of course --- it is meaningless on my system).
